docs/user-guide/a2-01-spirv-target-specific.md

```diff
@@ -0,0 +1,381 @@
+  {
+    v[i][j] = i * 4 + j;
+  }
+}
+```
+The Slang shader above can iterate each element of a `float3x4` matrix. This is similar to how a multi-dimensional array is handled in C and HLSL. When a matrix type is `float3x4`, the first dimension indexing, `i`, corresponds to the first value specified in the matrix type `3`. And the second dimension indexing, `j`, corresponds to the second value specified in the matrix type `4`.
+
+A matrix in Slang can be also seen as an array of a vector type. And the following code is same as above.
+```
+float3x4 v;
+for (int i = 0; i < 3; ++i)
+{
+  v[i] = float4(0, 1, 2, 3);
+  v[i] += i * 4;
+}
+```
+
+For the given example above, when targeting SPIR-V, Slang emits a matrix that consists of three vectors each of which has four elements,
+```
+%float = OpTypeFloat 32
+%v4float = OpTypeVector %float 4 ; <= float4 type
+%mat3v4float = OpTypeMatrix %v4float 3 ; <= three of float4
+```
+
+An alternative way to emit SPIR-V code is to emit four vectors and each vector has three elements. Slang doesn't do this but this is a more direct translation because SPIR-V spec defines OpTypeMatrix to take "Column Count" not row.
+```
+; NOT SLANG EMITTED CODE
+%v3float = OpTypeVector %float 3 ; <= float3 type
+%mat4v3float = OpTypeMatrix %v3float 4 ; <= four of float3
+```
+However, this results in a more complicated access pattern to the elements in a matrix, because `v[i]` will no longer correspond to a vector natively when emitted to SPIR-V.
+
+Another way to put, Slang treats column as row and row as column when targeting GLSL or SPIR-V. This is same to [how DXC handles a matrix when emitting SPIR-V](https://github.com/Microsoft/DirectXShaderCompiler/blob/main/docs/SPIR-V.rst#appendix-a-matrix-representation).
+
+Due to the swap of row and column in terminology, the matrix multiplication needs to be performed little differently. Slang translates a matrix multiplication, `mul(mat1, mat2)`, to `transpose(mul(transpose(mat2), transpose(mat1)))` when targeting SPIR-V.
+
+Note that the matrix translation explained above is orthogoal to the memory layout of a matrix. The memory layout is related to how CPU places matrix values in the memory and how GPU reads them. It is like how `std140` or `std430` works. DXC by default uses `column_major` memory layout and Slang uses row-major memory layout. For more information about the matrix memory layout, please see [a1-01-matrix-layout](a1-01-matrix-layout.md).
+
+
+Legalization
+------------
+
+Legalization is a process where Slang applies slightly different approach to translate the input Slang shader to the target.
+This process allows Slang shaders to be written in a syntax that SPIR-V may not be able to achieve natively.
+
+Slang allows to use opaque resource types as members of a struct.
+
+Slang allows functions that return any resource types as return type or `out` parameter as long as things are statically resolvable.
+
+Slang allows functions that return arrays.
```

Comment:
These functions will be converted to return the array via an `out` parameter in SPIRV.